src/app/data-model-provider/MetadataTypes.h

```diff
@@ -140,18 +167,26 @@ class ProviderMetadataTree
 public:
     virtual ~ProviderMetadataTree() = default;
 
-    virtual EndpointId FirstEndpoint()                 = 0;
-    virtual EndpointId NextEndpoint(EndpointId before) = 0;
+    virtual EndpointEntry FirstEndpoint()                              = 0;
+    virtual EndpointEntry NextEndpoint(EndpointId before)              = 0;
+    virtual std::optional<EndpointInfo> GetEndpointInfo(EndpointId id) = 0;
     virtual bool EndpointExists(EndpointId id);
 
     // This iteration describes device types registered on an endpoint
     virtual std::optional<DeviceTypeEntry> FirstDeviceType(EndpointId endpoint)                                  = 0;
     virtual std::optional<DeviceTypeEntry> NextDeviceType(EndpointId endpoint, const DeviceTypeEntry & previous) = 0;
 
-    // This iteration will list all clusters on a given endpoint
-    virtual ClusterEntry FirstCluster(EndpointId endpoint)                              = 0;
-    virtual ClusterEntry NextCluster(const ConcreteClusterPath & before)                = 0;
-    virtual std::optional<ClusterInfo> GetClusterInfo(const ConcreteClusterPath & path) = 0;
+    using SemanticTag = Clusters::Descriptor::Structs::SemanticTagStruct::Type;
+    virtual std::optional<SemanticTag> GetSemanticTagAtIndex(EndpointId endpoint, size_t index) = 0;
```

Comment:
throughout all the other APIs we tried to not have index-based iteration (even though this is what ember exposes). Could we use the same iteration style here instead (like first/next)?

Reply:
Done, used the iteration style here.